### ISSUE TITLE
No pinentry

### ISSUE THREAD

**Issue by chrBrd** [flathub/org.kde.kleopatra#56]
Issue as in title.

**Comment by travier** [comment]
This is not enough information to help you here.

**Comment by Erick555** [comment]
did you  have gpg-agent running as [instructions](https://github.com/flathub/org.kde.kleopatra#notice)?

**Comment by chrBrd** [comment]
> This is not enough information to help you here.

I'm afraid there's nothing I can add that's genuinely useful. `pinentry `is installed on the system, `gpg-agent` is running.

 I've even tried using `--pinentry-program` with a couple of different pinentry binaries when starting `gpg-agent`.
 
 Issue still as in title.

**Comment by Erick555** [comment]
Do you have gpg-agent socket in `/run/user/<id>/gnupg` ?

It was confirmed to work in flatpak and in kleopatra specifically

https://blog.grdryn.me/blog/flatpak-emacs-with-gpg-agent.html

https://github.com/flathub/flathub/pull/3182#issuecomment-1242964799

**Comment by chrBrd** [comment]
It looks like it only works with `pinentry-qt`, I've just got it working by installing that and starting the daemon with:
`gpg-agent --pinentry-program /usr/bin/pinentry-qt --daemon`.

**Comment by Keyaku** [comment]
I'm also running into this issue; using `pinentry-qt` as mentioned by @chrBrd solves it, which means I have to put it in `gpg.conf`.

**Note**: I'm currently using `socket=gpg-agent` instead of `host=xdg-run/gnupg`; using `host=xdg-run/gnupg` results in the same issue.

**Comment by Erick555** [comment]
Does it work agent was running before klepatra?

**Comment by Keyaku** [comment]
> Does it work when agent is running before klepatra?

Yes, but only with the pinentry option mentioned above.

**Comment by chrBrd** [comment]
If the agent isn't running before Kleopatra then there are no certificates to encrypt or decrypt anything with; pinentry isn't really relevant.

Things have changed for me after the recent Flatpak update and `pinentry-qt` no longer works, but console pinentry methods now do (with caveats, see below). 

I've done some brief experimenting with different pinentry options, starting Kleopatra using `flatpak run -vvvv` in a console.

**OS:** openSUSE Tumbleweed with SwayWM

**Kleopatra details:**
Version: 3.1.28.230805
Origin: flathub
Installation: system
Installed: 165.6 MB
Runtime: org.kde.Platform/x86_64/5.15-23.08
Sdk: org.kde.Sdk/x86_64/5.15-23.08
Commit: a9611316e6c10ca00ce0789e8916aafcfa5eba092646a2b68f8a9ffc3e5a125d

Error messages below are as shown in the 'Diagnostics' popup, eg:
```
gpg: signing failed: <ERROR_MSG>
gpg: -&38: sign+encrypt failed: <ERROR_MSG>
```

No additional information was shown in the terminal output from `flatpak run...` when attempting any of the below.

`pinentry-[qt/qt4/qt5/gtk2]`: '_No pinentry_'
`pinentry-gnome3`: '_pinentry error_'
`pinentry-[curses/tty/fltk]`: Works, but (on my system, might be different in a full DE) if Kleopatra isn't started from a console with `flatpak run ...` then the `pinentry-...` process will run in the background and ultimately timeout.

**Comment by gamer191** [comment]
It took me about 2 hours to figure out how to change the pinentry program kleopatra uses. For other users having this issue, the command `sudo update-alternatives --config pinentry` will interactively ask you which program should be the default pinentry method

Shoutout to https://unix.stackexchange.com/a/447045

**Comment by orionn333** [comment]
I noticed this some months ago, but it wasn't until now that I decided to look further into the matter.

I had already configured gpg-agent to be started on system startup:  
`gpg-agent --pinentry-program /usr/bin/pinentry-qt --daemon`

Checking the logs, I found the following:  
```
Failed to create wl_display (No such file or directory)
qt.qpa.xcb: could not connect to display :0
qt.qpa.plugin: From 6.5.0, xcb-cursor0 or libxcb-cursor0 is needed to load the Qt xcb platform plugin.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vkkhrdisplay, vnc, wayland-egl, wayland, xcb.
Process 20311 (pinentry-qt) of user 1000 terminated abnormally without generating a coredump.
```

So I decided to manually run pinentry-qt *(typing GETPIN)* and look up for problems in my host setup, but it ran successfully.

After replacing the pinentry program in gpg-agent with a custom one (so as I could inspect the env. vars.), I finally discovered what was wrong:  
pinentry-qt was being invoked with the env. var. `WAYLAND_DISPLAY=/run/flatpak/wayland-0`.  
I'm not super knowledgeable in Flatpaks or anything, so my first impression was that they don't have access to the host user wayland socket, which is expected.

When pinentry is called, the env. vars. are the ones from the Flatpak environment, and since gpg-agent is being run outside the Flatpak container, pinentry-qt cannot be launched (obviously, `/run/flatpak/wayland-0` does not exist on the host).

As a workaround, I created a wrapper that sets `WAYLAND_DISPLAY=wayland-0` and then calls pinentry-qt. I configured gpg-agent to use it, and now Kleopatra works as expected.  

This started happening on Flatpak [1.15.6](https://github.com/flatpak/flatpak/releases/tag/1.15.6), mainly due to:  
- Add support for Wayland security context ([#4920 ](https://github.com/flatpak/flatpak/pull/4920)).

Previous behaviour was:
    1. If the env. var. `WAYLAND_DISPLAY` wasn't set, it would then set its value to "wayland-0" (which is `/run/user/<user-id>/wayland-0`).  
    2. Create a symlink, inside the sandbox, located in the expected path, according to the value of `WAYLAND_DISPLAY` *(for "wayland-0",* `/run/user/<user-id>/wayland-0`*)*, and which would point to `/run/flatpak/<wayland-display>`, the sandboxed Wayland socket. *(In the prev. ex.:* `/run/flatpak/wayland-0`*)*.
    `WAYLAND_DISPLAY` **remains the same**.
    
But now:
    1. A new private Wayland socket is created in the host (`/run/user/<user-id>/.flatpak/wl/wayland-<XXXXXX>`), which is then binded to the sandbox path `/run/flatpak/<wayland-display>`.  
    2. Inside the sandbox, `WAYLAND_DISPLAY` value **is updated** to `/run/flatpak/<wayland-display>`.

I have not investigated further, but from what I've seen so far, if some random env. vars. and `WAYLAND_DISPLAY` are modified inside the sandbox, the `WAYLAND_DISPLAY` value inside the sandbox will be passed to pinentry while the other variables won't. The rest of env. variables that were provided to gpg-agent are the ones passed to pinentry, no matter what was done inside the sandbox. I guess it has something to do with bwrap, but I don't know for sure.

Technically, it will also work if the wrapper is modified to call `env WAYLAND_DISPLAY=/run/user/<user-id>/.flatpak/wl/wayland-<XXXXXX> pinentry-qt`, but the "XXXXXX" string MUST correspond with an active socket (ideally from the “parent” Kleopatra instance). There will be different sockets for every instance of any Flatpak app (at least from the current boot), so it isn't practical at all as a workaround.

### Wrapper

pinentry-wrapper.sh:  
```bash
#!/usr/bin/sh
env WAYLAND_DISPLAY=wayland-0 pinentry-qt "$@"
```

### TL;DR
With *Flatpak 1.15.6* (or higher), Kleopatra invokes pinentry-qt with the environment variable `WAYLAND_DISPLAY=/run/flatpak/<wayland-display>`, and since pinentry runs at host level instead of within the Flatpak sandbox, it **can't get access to a Wayland display**.  
I got Kleopatra to successfully retrieve data from pinentry-qt by creating a wrapper that sets `WAYLAND_DISPLAY=wayland-0` and then calls pinentry.

**Comment by Erick555** [comment]
Please report it as regression in https://github.com/flatpak/flatpak/issues

**Comment by ErikMichelson** [comment]
As a workaround you can use `pinentry-qt` when running from Flatpak (Kleopatra) and use `pinentry-gnome3` or whatever you prefer when running otherwise.
To achieve that, place a simple script like the following one into `/usr/local/bin/pinentry-wrapper` and configure that as your `pinentry-program`.

```bash
#!/bin/bash

if [[ "${DBUS_SESSION_BUS_ADDRESS}" == *"flatpak"* ]]; then
    /usr/bin/pinentry-qt "$@"
else
    /usr/bin/pinentry-gnome3 "$@"
fi
```

**Comment by Justinzobel** [comment]
Looks to have been fixed and pushed into main and I assume available in Flatpak 1.16.0, does anyone have this version and can test?

**Comment by Keyaku** [comment]
Here are the results of my testing:

Firstly, Kleopatra seems to be hardcoded to check `~/.gnupg`, which does not follow the XDG directory spec (should be `$XDG_DATA_HOME/gnupg`, or, in case of the app's manifest, `xdg-data/gnupg:create`). With this latter setting corrected, the logs show
```log
F: Not replacing "/home/USER/.gnupg" with tmpfs: Unable to open path "/home/USER/.gnupg": No such file or directory
```
obviously since 1. It's looking for a hardcoded directory in the sandbox, and 2. Even if it weren't hardcoded and looking in the user's home, the original `~/.gnupg` doesn't have that directory anyway (for people who follow the XDG spec at least), creating it would result in Kleopatra hosting its own files instead of looking for the user's currently installed certificates.

Secondly, not running the `pinentry` workarounds mentioned above, e.g. I have it on my `.zshenv` as:
```shell
if command -v gpg-agent &>/dev/null && ! killall -0 gpg-agent &>/dev/null; then
	gpg-agent --pinentry-program /usr/bin/pinentry-qt --daemon
fi
```
results in a longer app launch, followed by the main window of Kleopatra with the message "Loading certificate cache..." and a loading bar running indefinitely.

If the `pinentry` workaround is a requirement _for the user_ prior to launching Kleopatra (and not responsibility of the app), then the original issue seems to have been fixed; Kleopatra no longer loads an extra, unwanted/unconfigured `gpgconf` without a pinentry daemon present.
However, it would be best to make sure that:
1. The XDG spec is respected;
2. The users installing Kleopatra are at least informed that they need to run a `pinentry` daemon prior to launching the app.

**Comment by Erick555** [comment]
1. I'm not sure I understand [~/.gnupg](https://www.gnupg.org/documentation/manuals/gnupg/GPG-Configuration.html#index-_007e_002f_002egnupg) is default directory used for the past 25 years by gnupg itself. You may change it through `GNUPGHOME` variable but then you need to add permission for new path manually. If you want it to follow XDG standard by default then you need to ask in gnupg upstream.
2. Perhaps adding `pinentry-program /usr/bin/pinentry-qt` in `gpg-agent.conf` would make your life bit easier.

**Comment by Keyaku** [comment]
> 1. I'm not sure I understand [~/.gnupg](https://www.gnupg.org/documentation/manuals/gnupg/GPG-Configuration.html#index-_007e_002f_002egnupg) is default directory used for the past 25 years by gnupg itself.

I was so deep in "~/dotfile cleanup" land that I completely forgot about this. You are absolutely correct! I retract my statement (though I may take up on that offer of nudging about the XDG standard in gnupg upstream).
It's a shame we can't use environment variables when defining these filesystems in Flatpak apps, but that's another story.

> 2. Perhaps adding `pinentry-program /usr/bin/pinentry-qt` in `gpg-agent.conf` would make your life bit easier.

For some reason, this wouldn't work as expected in most of my systems; hence my surefire way from `.zshenv`. I might re-test it again, but regardless: this one isn't so much as making _my_ life easier, but everyone who installs Kleopatra.

**Comment by MikeNavy** [comment]
Hi,

Operating system: Linux Mint Mate 21.3 based on Ubuntu 22.04 LTS.
Flatpak Kleopatra Version: 4.0.0.250401 (25.04.1).
Runtime: org.kde.Platform/x86_64/6.8.
Flatpak framework version: Flatpak 1.16.1.

When trying to sign a document, "no pinentry error". 

I have on my operating system a full working GnuPG installation (GnuPG 4.6, libgcrypt 1.10.3, dirmngr, pinentry, gpg-agent...).

Issue might be related to this discussion "Pinentry with flatpak applications", see:
https://lists.gnupg.org/pipermail/gnupg-users/2025-February/067501.html

Letting pinentry outside the flatpak sandbox is, imo, an error: pinentry is a very sensitive component, used to enter passphrases without traces, and should be placed INSIDE the sandbox.

--> Without pinentry, flatpak Kleopatra is useless.

Note that Kleopatra 22.04.3 installed on my system from repositories DOES work, the problem is flatpak specific.

Regards,

MN

**Comment by Erick555** [comment]
The problem is gpg doesn't like when multiple instances are running in parallel so every app spawning its own daemon isn't proper solution. Flatpak protects host system from potentially malicious flatpak apps not apps from potentially malicious host so putting pinentry inside sandbox doesn't give you extra security.

**Comment by MikeNavy** [comment]
Hi,

At the moment flatpak Kleopatra suffers from two problems:

- It uses gpg binaries found in the runtime, org.kde.Platform version 
6.8, but these binaries are the ones of gpg 2.5.1, a test release and 
not a stable version. This has been corrected in the master of 
org.freedesktop.Platform (replacement by gpg 2.4.8), but we will have to 
wait for the use of org.kde.Platform version 6.10 by Kleopatra before to 
see this correction (in one year?).

- Since recent changes in flatpak framework or in the runtime, calling 
pinentry no longer works. This problem is a few months old, it affects 
all the flatpak applications using gpg, and it renders flatpak Kleopatra 
totally useless. At the moment, only some unofficial workarounds 
(changing gpg configuration on the user operating system) have been 
proposed.

- Flatpaks are sandboxes, and are there to make a total isolation 
between the sandboxed application and the operating system. A flatpak 
application is supposed to use its own binaries and the required 
dependencies found in the runtime. Here, flatpak Kleopatra uses its own 
binaries, gpg binaries found in the runtime, and pinentry from the 
operating system. This is an exception to flatpak principles and might 
be a source of vulnerability. Flatpak is based on two principles: the 
use of a sandboxing system (bubblewrap) and the use of the libraries and 
dependencies found in the runtime instead of system one. With that, if 
an instability occurs in the sandbox, it will not be propagated to the 
system, and the system staying in a stable state, no exploit will be 
possible outside the sandbox.

- Finally, I don't understand your answer "gpg doesn't like when 
multiple instances are running in parallel so every app spawning its own 
daemon isn't proper solution". A gpg instance fully running in the 
sandbox would be totally isolated from the operating system, and both 
gpg instances (in the operating system and in the flatpak sandbox) would 
not perceive the other one. Same thing, I don't understand your "Flatpak 
protects host system from potentially malicious flatpak apps not apps 
from potentially malicious host". Could you comment?

Regards,

MN

_PS_: I went back to Kleopatra 22.04.3 installed on my operating system 
from the "kubuntu backports" PPA, using gpg 2.4.6 from "rhansen gnupg2" 
PPA. I use Firejail sandboxing, profiles for gpg binaries exist (gpg, 
gpg-agent), and I have created a Kleopatra firejail profile by adapting 
the GPA profile provided with firejail (GPA is another GUI for gpg). 
Security is probably less than with flatpak, but higher than without 
sandboxing. And it does work...


Le 06/06/2025 à 19:07, Erick555 a écrit :

**Comment by Erick555** [comment]
>  It uses gpg binaries found in the runtime, org.kde.Platform version
> 6.8, but these binaries are the ones of gpg 2.5.1, a test release and
> not a stable version

yes, but is it an issue or just observation?

> Since recent changes in flatpak framework or in the runtime, calling
> pinentry no longer works

generally true but those are 4 year old [changes](https://github.com/flatpak/flatpak/commit/40510e8ae8fd433df04959938b8e81e582849481#diff-12ca338bd8d02696d700648c2b68705b2f354a228a4b6cbab5cd9e4335f63dabR317-R884)

> A flatpak
> application is supposed to use its own binaries and the required
> dependencies found in the runtime

We aren't talking about using host libraries but talking to host daemons like display server, audio server, smartcard manager and so on. gpg-agent is one of those daemons and `filesystem=xdg-run/gnupg` or it equivalent `socket=gpg-agent` flatpak permissions are specifically made to allow that and what flatpak apps are supposed to do which is consistent and intentional.

>  A gpg instance fully running in the
sandbox would be totally isolated from the operating system, and both
gpg instances (in the operating system and in the flatpak sandbox) would
not perceive the other one

Do you mean you have separate gpg keys storage for kleopatra and every other app on your system instead of shared `~/.gnupg` directory?

In general this isn't kleopatra issue so better to discuss either at gnupg thread you linked or https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/issues/1767

**Comment by MikeNavy** [comment]
Hi,

To summarize:

- At the moment, pinentry doesn't work in flatpak Kleopatra, and in all flatpak applications using GnuPG.

- Two issues have been closed at flatpak level, https://github.com/flatpak/flatpak/issues/5863. 

- On issue is still open at the freedesktop-sdk level, https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/issues/1767. As commented, "Flatpak fixed regression regarding wayland but no the dbus one". A patch has been proposed.

- GnuPG issued a code change in gpg-agent that addresses the problem. The patch needs to be applied to gpg-agent code, and gpg-agent recompiled.

The problem will not be solved until the proposed patch in freedesktop.sdk has been tested, approved, and ported to org.kde.Platform (in a few months...), or until a new version of gpg-agent is installed (on the operating system? in the flatpak runtime? in both?).

Meanwhile, several workarounds are possible:

- Use Kleopatra and GnuPG as available from your distribution repository.

- If Firejail is available for your distribution, install it (latest version is 0.9.74), and use the following profile for Kleopatra:

> include globals.local
noblacklist ${HOME}/.gnupg
noblacklist ${HOME}/.config
noblacklist ${HOME}/.local/share/kleopatra
include disable-common.inc
include disable-devel.inc
include disable-interpreters.inc
include disable-programs.inc
caps.drop all
netfilter
nodvd
nogroups
noinput
nonewprivs
noroot
nosound
notv
nou2f
novideo
protocol unix,inet,inet6
seccomp
tracelog
private-dev
restrict-namespaces

Copy this profile content, and save it as "kleopatra.profile" in "~/.config/firejail/". (I derived it from "gpa.profile" provided by Firejail; GPA is another GUI for GnuPG).
Fiirejail does provide profiles for "gpg" and "gpg-agent", that can be used "as is".
Finally, symlink "kleopatra", "gpg" and "gpg-agent" to firejail:

> sudo ln -s /usr/bin/firejail /usr/local/bin/kleopatra
sudo ln -s /usr/bin/firejail /usr/local/bin/gpg
sudo ln -s /usr/bin/firejail /usr/local/bin/gpg-agent

Once that done, "kleopatra", "gpg" and "gpg-agent" will run in Firejail sandboxes. With Firejail sandbox, security is enhanced, but isolation from the operating system is not total, since "kleopatra", "gpg" and "gpg-agent" will still use system dependencies. Security is lower than with Flatpak and higher than without sandboxing.

- For mail signatures and encryption only, you can use the flatpak version of Thunderbird: since version 78, Thunderbird includes its own OpenPGP code and does not rely any longer on GnuPG; of course, your keys need to be imported in Thunderbird.

Regards,

MN

**Comment by Erick555** [comment]
It's worth noting that the problem is specific to `pinentry-gnome3`. `pinentry-qt` should work out of the box so installing `pinentry-qt` and/or setting `pinentry-program /usr/bin/pinentry-qt` fixes the problem right away without squaring a wheel.
